### PR TITLE
Fixing apply(init) call for Transfrormer training

### PR DESCRIPTION
**Pull request by bfarzin**
I tried to train Transformer from scratch but was unable to make any progress in accuracy.  After researching a bunch of possible causes, it came down to the `init` method.  

Prior version would apply an `init` to the `Linear` layer that was tied to the `Embedding` layer for the inputs.  This put the mean to zero and std=0.02. Too small to learn anything.  

Two line change applies the `init` to the encoder half of the model (the Transformer part) and leaves the decoder alone.  I think this was the intention of passing in an `init` in the config for the model.  

A different decoder can train much faster for a Transformer type model (with either scaling our using attention.)  I am working on that now and will either put out as a PR or will post on forums for review first.

### CODE REVIEW

**Comment by sgugger** [comment]
I think this init method is the one that was used by default in the article introducing Transformer and used again in TransformerXL.
This PR won't initialize the decoder when tie_weights=False, and it doesn't change anything when tie_weight=True (since the decoder is tied to the encoder) so I don't see how it would change anything during training.

**Comment by bfarzin** [comment]
> This PR won't initialize the decoder when tie_weights=False, 

The decoder [does get an init](https://github.com/fastai/fastai/blob/6ba17b21599a6fc441794ffd130bc31b5333b4a0/fastai/text/models/awd_lstm.py#L140). Maybe we should have a better default in there?  `xavier_normal_` ?
I agree that if it is not tied, this uniform range might not be a good idea.

>it doesn't change anything when tie_weight=True (since the decoder is tied to the encoder) so I don't see how it would change anything during training.

That is what I thought too.  However, with a test, you can see how you link the `LinearDecoder` weights to the Embedding and then you now have a layer that is `Linear`.  Once you run the Init, you get a standard deviation of the weights that = 0.02 rather than 1.  And this breaks your ability to train (I am watching accuracy here, it does not move) An example follows:

```
from fastai.text import *
path = untar_data(URLs.IMDB_SAMPLE)
data_lm = TextLMDataBunch.from_csv(path, 'texts.csv')

config = tfmer_lm_config.copy()
config['n_layers'] = 6
config['n_heads'] = 6
config['ctx_len'] = 512
config['d_model'] = 768
config['d_inner'] = 3072

print(config)

model = get_language_model(Transformer, len(data_lm.vocab.itos), config=config, drop_mult=1.0)

print(f'encoder std: {model[0].encoder.weight.std()}')

learn = LanguageLearner(data_lm, model, **{'alpha':0,'beta':0})
learn.unfreeze()
learn.fit_one_cycle(5, max_lr=8e-5, div_factor=5, pct_start=1.0)
```
```
{'ctx_len': 512, 'n_layers': 6, 'n_heads': 6, 'd_model': 768, 'd_head': 64, 'd_inner': 3072, 'resid_p': 0.1, 'attn_p': 0.1, 'ff_p': 0.1, 'embed_p': 0.1, 'output_p': 0.0, 'bias': True, 'scale': True, 'act': <Activation.GeLU: 3>, 'double_drop': False, 'tie_weights': True, 'out_bias': False, 'init': <function init_transformer at 0x7fd609839378>, 'mask': True}
encoder std: 0.0200047567486763
epoch     train_loss  valid_loss  accuracy  time    
0         7.648284    6.933128    0.083482  00:27                                                                                   
1         6.992075    6.347262    0.083482  00:27                                                                                   
2         6.521496    5.987786    0.083482  00:27                                                                                   
3         6.254437    5.888292    0.083482  00:27                                                                                   
4         6.153402    5.879700    0.083482  00:27                                                                                   
5         6.127126    5.878162    0.083482  00:27                                                                                   
6         6.114800    5.875361    0.083482  00:27                                                                                   
7         6.112779    5.869525    0.083482  00:17                                                                                   
8         6.117646    5.877021    0.083482  00:12                                                                                   
9         6.116013    5.873543    0.083482  00:12                                                                                    
Total time: 03:56
```

Change the init, you can train:
```
nn.init.normal_(model[0].encoder.weight,0,1)
print(f'encoder std: {model[0].encoder.weight.std()}')
```
```
{'ctx_len': 512, 'n_layers': 6, 'n_heads': 6, 'd_model': 768, 'd_head': 64, 'd_inner': 3072, 'resid_p': 0.1, 'attn_p': 0.1, 'ff_p': 0.1, 'embed_p': 0.1, 'output_p': 0.0, 'bias': True, 'scale': True, 'act': <Activation.GeLU: 3>, 'double_drop': False, 'tie_weights': True, 'out_bias': False, 'init': <function init_transformer at 0x7f47b8008378>, 'mask': True}
encoder std: 0.01999562233686447
encoder std: 0.9997439384460449
epoch     train_loss  valid_loss  accuracy  time    
0         95.500992   40.837852   0.022351  00:12                                                                                    
1         53.763382   27.212849   0.091622  00:12                                                                                    
2         38.949074   21.942659   0.093318  00:12                                                                                    
3         31.039608   18.837091   0.088676  00:12                                                                                    
4         26.803869   21.505276   0.045521  00:12                                                                                    
5         24.380877   15.822821   0.095938  00:12                                                                                    
6         24.476906   23.919594   0.027321  00:12                                                                                    
7         23.012466   14.401671   0.117827  00:12                                                                                    
8         22.269203   19.510082   0.044673  00:12                                                                                    
9         19.240519   13.089678   0.107426  00:13 
```
One more tweak that I think is worth it for these models, you can scale the output and get better convergence (but I don't think this generalizes for all language models, or at least I have not tested that yet.)
```
from fastai.text import *                                                                                                                     
path = untar_data(URLs.IMDB_SAMPLE)                                                                                                           
data_lm = TextLMDataBunch.from_csv(path, 'texts.csv')                                                                                         
                                                                                                                                              
config = tfmer_lm_config.copy()                                                                                                               
config['n_layers'] = 6                                                                                                                        
config['n_heads'] = 6                                                                                                                         
config['ctx_len'] = 512                                                                                                                       
config['d_model'] = 768                                                                                                                       
config['d_inner'] = 3072                                                                                                                      
                                                                                                                                              
print(config)                                                                                                                                 
model = get_language_model(Transformer, len(data_lm.vocab.itos), config=config, drop_mult=1.0)                                                
                                                                                                                                              
class tfm_decoder(nn.Module):                                                                                                                 
    def __init__(self,d_model=768,d_vocab=8842, enc:nn.Module=None):                                                                          
        super().__init__()                                                                                                                    
        self.tgt_word_prj = nn.Linear(d_model, d_vocab, bias=False)                                                                           
        nn.init.xavier_normal_(self.tgt_word_prj.weight)                                                                                      
                                                                                                                                              
        if enc: self.tgt_word_prj.weight = enc.weight                                                                                         
        self.x_logit_scale = (d_model ** -0.5)                                                                                                
                                                                                                                                              
    def forward(self,input:Tuple[Tensor,Tensor])->Tuple[Tensor,Tensor,Tensor]:                                                                
        raw_outputs,outputs = input                                                                                                           
        decoded = self.tgt_word_prj(raw_outputs[0]) * self.x_logit_scale                                                                      
        return decoded, raw_outputs, outputs                                                                                                  
                                                                                                                                              
model[1] = tfm_decoder(768, len(data_lm.vocab.itos), model[0].encoder)                                                                        
                                                                                                                                              
print(f'encoder std: {model[0].encoder.weight.std()}')                                                                                        
nn.init.normal_(model[0].encoder.weight,0,1)                                                                                                  
print(f'encoder std: {model[0].encoder.weight.std()}')                                                                                        
                                                                                                                                              
learn = LanguageLearner(data_lm, model, **{'alpha':0,'beta':0})                                                                               
learn.unfreeze()                                                                                                                              
learn.fit_one_cycle(10, max_lr=8e-5, div_factor=5, pct_start=1.0) 
```
```
{'ctx_len': 512, 'n_layers': 6, 'n_heads': 6, 'd_model': 768, 'd_head': 64, 'd_inner': 3072, 'resid_p': 0.1, 'attn_p': 0.1, 'ff_p': 0.1, 'embed_p': 0.1, 'output_p': 0.0, 'bias': True, 'scale': True, 'act': <Activation.GeLU: 3>, 'double_drop': False, 'tie_weights': True, 'out_bias': False, 'init': <function init_transformer at 0x7f24c0af1400>, 'mask': True}
encoder std: 0.020003292709589005
encoder std: 0.9999027252197266
epoch     train_loss  valid_loss  accuracy  time    
0         8.145821    7.105320    0.080833  00:12                                                                                   
1         7.323191    6.781451    0.106726  00:12                                                                                   
2         7.014435    6.603555    0.122515  00:12                                                                                   
3         6.771075    6.363043    0.143080  00:12                                                                                   
4         6.533041    6.146296    0.155923  00:13                                                                                   
5         6.338411    5.988508    0.163185  00:13                                                                                   
6         6.185056    5.884889    0.169807  00:13                                                                                   
7         6.058295    5.779028    0.174539  00:12                                                                                   
8         5.933101    5.657488    0.183780  00:12                                                                                   
9         5.818246    5.565749    0.189628  00:12 
```
